FsAutoComplete/CompilerServiceInterface.fs

```diff
@@ -0,0 +1,159 @@
+namespace FsAutoComplete
+
+open System
+open System.IO
+open Microsoft.FSharp.Compiler
+open Microsoft.FSharp.Compiler.SourceCodeServices
+
+type ParseAndCheckResults(parse: FSharpParseFileResults,
+                          check: FSharpCheckFileResults,
+                          version: int) =
+  let parseResults = parse
+  let checkResults = check
+  let _version = version
+
+  member x.TryGetMethodOverrides (lines: string[]) (line: int) (col: int) =
+    // Find the starting point, ideally right after the first '('
+    let lineCutoff = line - 3
+    let commas, line, col =
+      let rec prevPos (line,col) =
+        match line, col with
+        | 1, 1
+        | _ when line < lineCutoff -> 1, 1
+        | _, 1 ->
+           let prevLine = lines.[line - 2]
+           if prevLine.Length = 0 then prevPos(line-1, 1)
+           else line - 1, prevLine.Length
+        | _    -> line, col - 1
+
+      let rec loop commas depth (line, col) =
+        if (line,col) <= (1,1) then (0, line, col) else
+        let ch = lines.[line - 1].[col - 1]
+        let commas = if depth = 0 && ch = ',' then commas + 1 else commas
+        if (ch = '(' || ch = '{' || ch = '[') && depth > 0 then loop commas (depth - 1) (prevPos (line,col))
+        elif ch = ')' || ch = '}' || ch = ']' then loop commas (depth + 1) (prevPos (line,col))
+        elif ch = '(' || ch = '<' then commas, line, col
+        else loop commas depth (prevPos (line,col))
+      match loop 0 0 (prevPos(line,col)) with
```

Comment:
It's a shame, because really the editor knows the most about what in the file has changed, and typically knows whether we are in a string or a comment, but we don't want to push that out to multiple implementations. To have the full state of the file tokenized would really require sending deltas to FsAutoComplete, which would work well, but probably be very fiddly to implement for each editor.

Let's assume for the sake of argument that the editor knows not to make requests in the middle of strings or comments. Then the problem is to look backwards far enough to find a good starting point to tokenize up to where we are now. I wonder if we could leverage the significant whitespace to look backwards for the start of the current block. Unfortunately a simple dedent is not enough, as the following is legal:

```fsharp
let s = String.Format("hello",
                  "good",
                      2
                          )
```